### ISSUE TITLE
fast-forward tokens in logits post processor

### ISSUE THREAD

**Issue by mmoskal** [NVIDIA/TensorRT-LLM#2365]
I've been working on an [OpenAI-compatible REST server](https://github.com/guidance-ai/llgtrt), utilizing TensorRT-LLM but not Triton, similar to `openai_server.py` but in Rust and generally production-ready. Its main feature is support for constrained decoding, using [Guidance](https://github.com/guidance-ai/guidance) or more specifically the Rust [low-level Guidance library](https://github.com/microsoft/llguidance). This allows for enforcing JSON schemas on the output (similar to OpenAI's "Structured Output" feature, though much more general and with no initial latency), as well as arbitrary context-free grammars.

I've been utilizing the batched logits post processor - I wrote a [cuda kernel](https://github.com/guidance-ai/llgtrt/blob/main/trtllm-c/mask_logits.cu) that applies boolean mask to logits (ie., sets logits to -inf where mask is false). Because I needed to support different temperature per token, the kernel also scales logits. There is an additional hack in there to support temperature==0.0 (greedy sampling). It would be nice it logits post processor allowed for changing sampling parameters.

However, the main feature that is missing and that would help performance is the ability to append more than one token to the sequence in the output phase - fast-forward tokens. This can be thought of an "speculation that is always right".

An example, where fast forward tokens are useful is generating data adhering to a certain JSON schema. The logit processor forces `{"name":"` to be generated, then the model generates `John"`, the processor forces `,\n"age":`, model generates `42`, and so on. Another example is chain-of-thought reasoning, where after the model generated a sentence, the controller forces more instructions for the model, the model generates more text, and so on. If used, these greatly speed up generation process.

The sglang [implements this](https://lmsys.org/blog/2024-02-05-compressed-fsm/) though it was [first done](https://github.com/guidance-ai/guidance/discussions/138) in Guidance.

One way to allow this functionality is to extend logit post-processor with the following callback that for each sequence is given the sampled token and returns a vector of tokens (typically it would be just the sampled token in the vector but could be something else, not necessarily starting with the sampled token).

```cpp
using PostSamplingCallback = std::function<   std::vector<std::vector<std::vector<TokenIdType>>> (std::vector<IdType> const &seq_ids,
    std::vector<std::vector<TokenIdType>> const &sampled_tokens, 
    std::vector<std::optional<IdType>> const &user_seq_ids)>;
```

The callback would take request ids, sampled tokens - a vector of num_requests of vectors of beam size of sampled tokens, and user req ids. It would return a vector of num_requests, of vectors of beam size of vectors of tokens to append to each beam.

For example:
- model generates `John`, post-processor returns {`John`}
- model generates `"`, post-processor returns {`",`, `\n`, `"`, `age`, `": `} (assuming such tokens exists - note how `"` was replaced by `",`)

The signature above looks somewhat complicated due to beams and requests, for a single beam, single request it's essentially:

```cpp
vector<Token> post_sampling(Token t) { 
  return {t}; // default
}
```

**Comment by MartinMarciniszyn** [comment]
Hi @mmoskal, thank you for submitting this suggestion. We are in the process of improving the capabilities for generating structured output in TRT-LLM. We shall take your proposal into consideration. I will get back to you once we have a clearer design. CC @juney-nvidia , @tqchen

**Comment by mmoskal** [comment]
Thank you @MartinMarciniszyn ! If you are interested in integrating the llguidance library directly into TRT-LLM let me know. The library has a [C interface](https://github.com/microsoft/llguidance/blob/main/parser/llguidance.h) that allows for regex, JSON schema, and Guidance constraints.

**Comment by lupickup** [comment]
@MartinMarciniszyn Adding to the interest in enabling fast-forward tokens. Azure AI has been collaborating with @mmoskal and we're very interested in combining the best-in-class inference performance of trt-llm with the best-in-class constrained decoding features guidance and it's ecosystem of libraries offers!

**Comment by mmoskal** [comment]
We have just released a large [JSON Schema benchmark](https://github.com/guidance-ai/jsonschemabench) and a [paper](https://arxiv.org/abs/2501.10868). The paper in particular has data on how much token can be typically fast-forwarded (see difference between LLM-only and Guidance in the plot below)

![Image](https://github.com/user-attachments/assets/16680db7-1f85-4fc6-a8d2-fbfb33277b54)

There is also [isolated mask-generation benchmarks](https://github.com/guidance-ai/jsonschemabench/tree/main/maskbench) - comparing LLGuidance, Outlines, XGrammar and llama.cpp grammars.

**Comment by MartinMarciniszyn** [comment]
Thank you for sharing the benchmark.

Note that XGrammer has been integrated into TRT-LLM in the meanwhile. [Here is an example](../blob/main/examples/llm-api/llm_guided_decoding.py).

We are in the process of evaluating how to implement the ability of appending more than one token to the sequence. Instead of extending the interface of `LogitsPostprocesser`, one could also follow a target-draft model approach or a modified version of lookahead decoding. We will share updates here.

**Comment by mmoskal** [comment]
I love the integration of constrained decoding directly into TRT-LLM!

But, I would like to see LLGuidance integrated as well, as it's significantly faster than XGrammar and has much more JSON feature coverage (eg. `allOf`, `min/maxItems`, `minimum/maximum`, `multipleOf`, `additionalProperties`, sibling keys; note that implementing some of these makes the grammar more complex, which is why "XGrammar.cpp" which is using llama.cpp JSON->Grammar converter is so much slower than regular XGrammar).

![Image](https://github.com/user-attachments/assets/6a6c434a-ecf2-49ed-acd2-8c3272615310)

See [MaskBench](https://github.com/guidance-ai/jsonschemabench/tree/main/maskbench) for details.

**Comment by karljang** [comment]
Closing this issue as stale. 
If you're still exploring speedups, consider trying guided decoding combined with speculative decoding, they may offer similar performance improvements.
Thank you!
